src/librustc/mir/transform/dataflow.rs

```diff
@@ -0,0 +1,419 @@
+use mir::repr as mir;
+use mir::cfg::CFG;
+use mir::repr::BasicBlock;
+use rustc_data_structures::bitvec::BitVector;
+
+use mir::transform::lattice::Lattice;
+
+
+pub trait DataflowPass<'tcx> {
```

Comment:
I'm not sure that `DataflowPass` is pulling its weight. `ar_forward` already takes two parameters, one for the analysis and one for the pass. If `DataflowPass` was removed and `ar_forward` took two parameters, one for analysis, and one for rewriting, that would require less boilerplate while being just as ergonomic. Moreover, I don't think it really makes sense to tie the transfer function to the rewrite - as an example, no rewrite cares how alias results are generated, but there are many different alias analyses out there, so rewrites should not be tied to whatever alias analysis is used.

This would require that `Transfer` have its `Lattice` as an associated type instead of a parameter, but I can't see an example where it makes sense to implement `Transfer` for different lattices.